src/Lib/Modes/TransformMode.vala

```diff
@@ -24,57 +24,433 @@
  * be used to apply the underlying code on top of other modes that may need to
  * use the functionality.
  */
-public class Akira.Lib.Modes.TransformMode : Object, InteractionMode {
+public class Akira.Lib.Modes.TransformMode : InteractionMode {
+    private const double ROTATION_FIXED_STEP = 15.0;
+
     public weak Akira.Lib.Canvas canvas { get; construct; }
     public weak Akira.Lib.Managers.ModeManager mode_manager { get; construct; }
 
-    public TransformMode (Akira.Lib.Canvas canvas, Akira.Lib.Managers.ModeManager mode_manager) {
+    public class InitialDragState {
+        public double press_x = 0.0;
+        public double press_y = 0.0;
+        public double nob_x = 0.0;
+        public double nob_y = 0.0;
+
+        public double item_x = 0.0;
+        public double item_y = 0.0;
+        public double item_width = 0.0;
+        public double item_height = 0.0;
+        public Cairo.Matrix item_transform;
+
+        public double item_scale_x_adj = 0.0;
+        public double item_scale_y_adj = 0.0;
+
+        public double rotation_center_x = 0.0;
+        public double rotation_center_y = 0.0;
+
+        public bool wants_snapping = true;
+    }
+
+    public class TransformExtraContext : Object {
+        public Akira.Lib.Managers.SnapManager.SnapGuideData snap_guide_data;
+    }
+
+    public InitialDragState initial_drag_state;
+    public TransformExtraContext transform_extra_context;
+
+
+    public TransformMode (Akira.Lib.Canvas canvas, Akira.Lib.Managers.ModeManager? mode_manager) {
         Object (
             canvas: canvas,
             mode_manager : mode_manager
         );
     }
 
-    public void mode_begin () {}
-    public void mode_end () {
+    construct {
+        initial_drag_state = new InitialDragState ();
+        transform_extra_context = new TransformExtraContext ();
+        transform_extra_context.snap_guide_data = new Akira.Lib.Managers.SnapManager.SnapGuideData ();
+    }
+
+    public override void mode_begin () {
+        unowned var selected_items = canvas.selected_bound_manager.selected_items;
+        if (!initialize_items_drag_state (selected_items, ref initial_drag_state)) {
+            debug ("TransformMode only works if an item is selected");
+            if (mode_manager != null) {
+                mode_manager.deregister_mode (mode_type ());
+                return;
+            }
+        }
+    }
```

Comment:
I'm not a fan of indented conditions.
Would be possible to revert the conditions and use early returns to keep 1 level of indentation?